File: SpringBoot-App/src/main/java/tripkovic/andrej/pmu/model/RoadProblem.java
package tripkovic.andrej.pmu.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.Data;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

@Data
@Document(collection = "road_problems")
public class RoadProblem {

    @Id
    private String id;

    private double latitude;
    private double longitude;

    private String title;
    private String description;

    private String type;      // WORK_ON_ROAD, PROBLEM_ON_ROAD, OTHER
    private String severity;  // LOW, MEDIUM, HIGH, CANT_PASS

    private String roadSide = "BOTH"; // BOTH, MY_SIDE, OPPOSITE — relative to directionBearing
    private Double directionBearing;  // reporter's travel bearing in degrees (0–360), null if unknown

    private boolean userReport;
    private boolean official = false; // reported by a moderator or system-seeded
    private String reportedBy;
    private String photoBase64;

    private boolean hidden = false;
    private int upvotes = 0;
    private int downvotes = 0;

    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
    private LocalDateTime createdAt = LocalDateTime.now();
}

File: Android-App/app/src/main/java/com/example/myapplicationtest/model/RoadProblem.kt
package com.example.myapplicationtest.model

import android.net.Uri
import com.google.android.gms.maps.model.LatLng

data class RoadProblem(
    val id: String? = null,
    val position: LatLng,
    val title: String,
    val description: String,
    val type: RoadProblemType,
    val severity: Severity = Severity.MEDIUM,
    val roadSide: RoadSide = RoadSide.BOTH,
    val directionBearing: Double? = null,
    val photoUri: Uri? = null,
    val photoBase64: String? = null,
    val isUserReport: Boolean = false,
    val official: Boolean = false,
    val reportedBy: String? = null,
    val upvotes: Int = 0,
    val downvotes: Int = 0,
    val createdAt: String? = null,
    val hidden: Boolean = false,
) {
    /**
     * Bearing (degrees, 0–360) of the travel direction this problem affects,
     * or null if it affects both directions (or the direction is unknown).
     */
    val affectedBearing: Double?
        get() = when (roadSide) {
            RoadSide.BOTH -> null
            RoadSide.MY_SIDE -> directionBearing?.let { (it % 360.0 + 360.0) % 360.0 }
            RoadSide.OPPOSITE -> directionBearing?.let { ((it + 180.0) % 360.0 + 360.0) % 360.0 }
        }
}
